Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py

```diff
@@ -0,0 +1,28 @@
+import FWCore.ParameterSet.Config as cms
+
+import HLTrigger.HLTfilters.hltHighLevel_cfi as hlt
+
+ALCARECOPPSCalMaxTracksFilter = hlt.hltHighLevel.clone(
+  TriggerResultsTag = cms.InputTag("TriggerResults","","HLTX"),
+  eventSetupPathsKey = 'PPSCalMaxTracks',
+)
+
+from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import *
+
+totemRPRawToDigi.rawDataTag = cms.InputTag('hltPPSCalibrationRaw')
```

Comment:
Shouldn't these be cloned and renamed in order not to change the parameters of the original?

Reply:
If I do that, for example, replace the corresponding lines with:
```python
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemRPRawToDigi as _totemRPRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsDiamondRawToDigi as _ctppsDiamondRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTimingRawToDigi as _totemTimingRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsPixelDigis as _ctppsPixelDigis

totemRPRawToDigi = _totemRPRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
ctppsDiamondRawToDigi = _ctppsDiamondRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
totemTimingRawToDigi = _totemTimingRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
ctppsPixelDigis = _ctppsPixelDigis.clone(inputLabel = 'hltPPSCalibrationRaw')
```

Then after running the test with:
```shell
# I've also tried simpler tests without L1Reco stage, namely:
cmsDriver.py testExpressPPSAlCaRecoProducer  -s ALCAPRODUCER:PPSCalMaxTracks,ENDJOB \
--process ALCARECO \
--scenario pp \
--era ctpps_2018 \
--conditions auto:run3_data_express \
--data  \
--datatier ALCARECO \
--eventcontent ALCARECO \
-n 100  --filein file:/eos/project-c/ctpps/subsystems/Software/Off-line/AlCaTest/outputALCAPPS_single.root \
--fileout file:outputALCAPPS_RECO_express.root \
--customise_commands="process.GlobalTag.toGet = cms.VPSet()\nprocess.GlobalTag.toGet.append(cms.PSet(record = cms.string(\"AlCaRecoTriggerBitsRcd\"),tag =  cms.string(\"AlCaRecoHLTpaths_PPS2022_express_v1\"), connect = cms.string(\"frontier://FrontierProd/CMS_CONDITIONS\")))"
```

I get the following error:
```
----- Begin Fatal Exception 02-Feb-2022 21:46:59 CET-----------------------
An exception of category 'NoRecord' occurred while
   [0] Processing  Event run: 322022 lumi: 1025 event: 1760583922 stream: 0
   [1] Running path 'ALCARECOoutput_step'
   [2] Prefetching for module PoolOutputModule/'ALCARECOoutput'
   [3] Calling method for module TotemVFATRawToDigi/'ctppsDiamondRawToDigi'
Exception Message:
No "TotemReadoutRcd" record found in the EventSetup.

 Please add an ESSource or ESProducer that delivers such a record.
----- End Fatal Exception -------------------------------------------------
```

Reply:
I think in order to have the `ESSource` executed this file should be loaded `EventFilter/CTPPSRawToDigi/python/ctppsRawToDigi_cff.py`

Reply:
I mean you can still import everything but then clone and rename only the stuff that gets modified.

Reply:
I've tried your suggestion, i.e. to import everything (first line `from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import *`) but then clone and rename only the stuff that gets modified (following lines):

```python
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import *
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemRPRawToDigi as _totemRPRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsDiamondRawToDigi as _ctppsDiamondRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTimingRawToDigi as _totemTimingRawToDigi
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsPixelDigis as _ctppsPixelDigis

totemRPRawToDigi = _totemRPRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
ctppsDiamondRawToDigi = _ctppsDiamondRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
totemTimingRawToDigi = _totemTimingRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
ctppsPixelDigis = _ctppsPixelDigis.clone(inputLabel = 'hltPPSCalibrationRaw')

ctppsRawToDigiTask = cms.Task(
  totemRPRawToDigi,
  ctppsDiamondRawToDigi,
  totemTimingRawToDigi,
  ctppsPixelDigis
)
```

Then after running the test with:
```bash
customise_commands="process.GlobalTag.toGet = cms.VPSet()\n\
process.GlobalTag.toGet.append(cms.PSet(record = cms.string(\"AlCaRecoTriggerBitsRcd\"),tag =  cms.string(\"AlCaRecoHLTpaths_PPS2022_prompt_v1\"), connect = cms.string(\"frontier://FrontierProd/CMS_CONDITIONS\")))\
\nprocess.ALCARECOPPSCalMaxTracksFilter.TriggerResultsTag = cms.InputTag(\"TriggerResults\",\"\",\"HLTX\")"

# note we currently use `auto:run3_data_express` GT 
# the correct GT (auto:run3_data_prompt) doesn't have LHCInfo record for run 322022 which corresponds to our face ALCARAW file
cmsDriver.py testPromptPPSAlCaRecoProducer  -s ALCAPRODUCER:PPSCalMaxTracks,ENDJOB \
--process ALCARECO \
--scenario pp \
--era ctpps_2018 \
--conditions auto:run3_data_express \
--data  \
--datatier ALCARECO \
--eventcontent ALCARECO \
-n 100  --filein file:/eos/project-c/ctpps/subsystems/Software/Off-line/AlCaTest/outputALCAPPS_single.root \
--fileout file:outputALCAPPS_RECO_prompt.root \
--customise_commands="$customise_commands"
```

I get the following error:
```python
    raise RuntimeError("module has no label.  Perhaps it wasn't inserted into the process?")
RuntimeError: module has no label.  Perhaps it wasn't inserted into the process?
(...)
    raise RuntimeError("An entry in sequence "+label + ' has no label')
RuntimeError: An entry in sequence ctppsRawToDigi has no label
```

Reply:
It seems that the problem is related to the fact that the failing relval tests (like `1001.2`) are loading into `process` the file `Configuration.StandardSequences.RawToDigi_Data_cff`.
This in turn imports `Configuration.StandardSequences.RawToDigi_cff` which calls:
```
from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import *
```
See https://github.com/cms-sw/cmssw/blob/CMSSW_12_3_X/Configuration/StandardSequences/python/RawToDigi_cff.py#L46

This loads into global namespace variables like: `totemRPRawToDigi`, `ctppsDiamondRawToDigi` etc...

After loading into `process` the file `Configuration.StandardSequences.RawToDigi_Data_cff` relval tests configuration proceeds to load into process the file `Configuration.StandardSequences.AlCaRecoStreams_cff`.
This, in turn, calls the import of our new AlCa reco producer `from Calibration.PPSAlCaRecoProducer.ALCARECOPPSCalMaxTracks_cff import *`.
Unfortunately in the `ALCARECOPPSCalMaxTracks_cff` the variables `totemRPRawToDigi`, `ctppsDiamondRawToDigi` are modified and redefined with custom `rawDataTag`.

In this way configuration of the new AlCa reco producer is poisoning the default RAW2DIGI step of PPS which is used in other relvals.

We have already tried various methods of importing in order to avoid such a situation but this has failed in many ways as seen in the discussion here: https://github.com/cms-sw/cmssw/pull/36273#discussion_r791759792

Reply:
@makortel @mmusich  do you have any suggestions here? Unfortunately I'm not sure what else could be done

Reply:
At the very least one could import here all the pieces that need to be loaded to run, but I am fairly certain there should be a way of reordering the includes to avoid framework issues. I might be mistaken but I think the current implementation is risky and probably even prohibited by coding rules.

Reply:
This is exactly the problem the policy of "EDModules are allowed to be modified only in the `.py` file they are defined" tries to prevent.

What should be done here is along
* clone the EDModules defined in `EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff` to different labels (i.e. variable names)
* if there are any downstream EDModules that intend to use the data products of the EDModules cloned here and are defined elsewhere, those would need to be cloned and customized as well

Reply:
@makortel what we are adjusting here are PPS modules responsible for RAW2DIGI. We need to adjust the labels of raw data to such ones that we expect to have in ALCARAW file (namely `hltPPSCalibrationRaw`). There are several modules downstream which use products of the modified RAW2DIGI module (i.e. modules doing RECO step for PPS subsystem).
Probably these modifications may go need to be propagated even further as the output of this ALCA RECO producer will be used in PCL loops.

Reply:
If this `cff` would never have to be included in a job running the "standard versions" of these EDModules, just overriding the parameters of an EDModule defined elsewhere could be tolerated (although even then it would be better to clone the existing EDModule into the same label, because then an attempt to use the `cff` fragment together with the standard version would result in an error instead of unexpected results).

If the `cff` needs to run along the standard versions (or it is uncertain), the modules that need customization must be duplicated. Framework contains some helpers to replace InputTags of modules in a Sequence
https://github.com/cms-sw/cmssw/blob/master/FWCore/ParameterSet/python/MassReplace.py
and PAT tools have helpers to clone entire Sequences
https://github.com/cms-sw/cmssw/blob/e5a269d0263d47bdf9ede055087cf470913f46f1/PhysicsTools/PatAlgos/python/tools/helpers.py#L257-L264

@cms-sw/alca-l2 Do any existing AlCaReco fragment do anything similar that could be used as an example?

Reply:
It seems to me these are doing in a nutshell the same thing you do here,  but for HCal: run raw2digi and then a modified reconstruction sequence:
  * https://github.com/cms-sw/cmssw/blob/master/Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalPedestal_cff.py
  * https://github.com/cms-sw/cmssw/blob/master/Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalMinBiasNoise_cff.py

Reply:
In the files @mmusich mentioned in https://github.com/cms-sw/cmssw/pull/36273#discussion_r798932994 the label changes had to be propagated into all modules:
1. Custom input tag `hltHcalCalibrationRaw` was added to the `hcalDigis` producer imported from (`EventFilter.HcalRawToDigi.HcalRawToDigi_cfi`). `hcalDigis` is cloned under a new name `hcalDigiAlCaPedestal`.
2. As digis are produced by `hcalDigiAlCaPedestal` and not by `hcalDigis`, the following modules have to be imported under new names
    1. `hfsimplereco` -> `hfrecoPedestal`
    2. `hosimplereco` -> `horecoPedestal`
    3. `hfprereco` -> `hfprerecoPedestal `
3. The change mentioned above triggered another renaming in modules that had to use new labels:
   1.  `hfreco` -> `_phase1_hfrecoPedestal `  (due to change `hfprereco` -> `hfprerecoPedestal `)
   2. `hbheplan1` -> `hbheplan1Pedestal `   (due to change `hfprereco` -> `hfprerecoPedestal `)

As we could see the simple change of one input label (to `hltHcalCalibrationRaw`) proliferates into more than 100 lines of the Python code. I was curious if there is a more elegant way of adjusting the label of a single module without the need of inspecting all dependencies and adjust all consumer modules.

I can try this approach for PPS but there is one more step needed - we would need to adapt all our PCL modules which would use data generated by this ALCA Reco producer and point them to new labels. This has to be tested as well.